core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -881,7 +881,7 @@ public EnvVars getEnvironment(TaskListener log) throws IOException, InterruptedE
                 e.buildEnvVars(env);
 
         for (EnvironmentContributingAction a : getActions(EnvironmentContributingAction.class))
-            a.buildEnvVars(this,env);
+            a.buildEnvVars(this,env,getBuiltOn());
```

Comment:
AFAICT this loop should just be deleted since it is already being called in the super; otherwise we would be calling each `EnvironmentContributingAction` twice on a single `AbstractBuild`.